### DISCUSSION TITLE
Is there a way to run an action step without networking?

### DISCUSSION THREAD

**Discussion by kinnison** [community/community#24975] [Actions]
<p>I would like to run a particular step of my workflow without access to networking (well without Internet access specifically).</p>
<p>This is a testing step and recently we had an issue where our tests were sometimes not being fully mocked and accidentally reached out to the Internet which is not ideal for reproducibility etc.</p>
<p>I would like to make it so that the testing step of our action is somehow blocked from accessing the Internet so that this kind of mistake is caught during PR review.</p>
<p>Does anyone have any recommendations of how to do that?</p>
<p>Thanks,</p>
<p>Daniel.</p>

**Comment by airtower-luna** [comment]
<p>My first thought: Run it in a container with networking disabled, e.g.:</p>
<pre><code class="lang-bash">docker run --network=none ...
</code></pre>

**Comment by kinnison** [comment]
<p>Is there a way to run in a ‘container’ which is just the host OS?  I’d rather not complicate matters with another “platform” if I can avoid it.</p>

**Comment by kinnison** [comment]
<p>Also, will that kind of thing work on Windows and Mac runners too?</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="kinnison" data-post="3" data-topic="162191" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 kinnison:</div>
<blockquote>
<p>Is there a way to run in a ‘container’ which is just the host OS? I’d rather not complicate matters with another “platform” if I can avoid it.</p>
</blockquote>
</aside>
<p>On Linux you could achieve that with <code>unshare --net ...</code>, to run a command in a network namespace of its own without starting a full container that’d also be separated in other ways (like file system).</p>
<p>I do something similar here, to run web server tests in separate network and IPC namespaces:</p><aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/airtower-luna/mod_gnutls/blob/461fd384dbc08afc2fd1bd1b66cb357bfa074312/test/netns_py.bash#L21-L22" target="_blank" rel="noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/airtower-luna/mod_gnutls/blob/461fd384dbc08afc2fd1bd1b66cb357bfa074312/test/netns_py.bash#L21-L22" target="_blank" rel="noopener">airtower-luna/mod_gnutls/blob/461fd384dbc08afc2fd1bd1b66cb357bfa074312/test/netns_py.bash#L21-L22</a></h4>
<pre class="onebox"><code class="lang-bash"><ol class="start lines" start="21" style="counter-reset: li-counter 20 ;">
<li>    exec "${UNSHARE}" --net --ipc -r /bin/bash -c \</li>
<li>	 "export MGS_NETNS_ACTIVE=1; ip link set up lo; exec ${UNSHARE} --user ${PYTHON} -u ${*}"</li>
</ol></code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>I don’t know if there’s anything similar for Windows or Mac.</p>

**Comment by kinnison** [comment]
<p>Ooh <code>unshare</code> is definitely an interesting option, I could look at doing that on Linux for sure.  My hope was there’d be a <code>networking: false</code> or somesuch for the shell step, but I guess that’s not something others have asked for :smiley:</p>

**Comment by chrispat** [comment]
<p>A container is going to be the only option here as that would be isolated from the host.  Disabling networking on the host during a step would cause the runner to lose connection to <a href="http://Github.com" rel="noopener nofollow ugc">Github.com</a> and your job would fail.</p>

**Comment by kinnison** [comment]
<p><a class="mention" href="/chrispat">@chrispat</a> Thanks for that input, it’s useful to prevent me trying down that route for sure.  Is there any way that a container / generic solution could be worked out which the runner could enforce on the processes used in the step, without limiting the runner itself, across all the platforms?  I figure this is getting into the realms of ‘not easy to make cross-platform’ but if anyone know what the runners are capable of, it’ll be someone like you :slight_smile: .</p>

**Comment by chrispat** [comment]
<p>I don’t think it is possible in a cross platform way.  The only model I can think of where this could be possible in a straight forward way would be using a container.  Maybe there is something you can do with HyperKit or virtual box on MacOS.</p>

**Comment by kinnison** [comment]
<p>Okay, thanks Chris.  I’ll see what I can get up to with unshare in my Linux builds and trust that my other platform’s code won’t vary too much :slight_smile:</p>

**Comment by theophanemayaud** [comment]
Funny, for GitHub copilot coding agent, that’s exactly what GitHub does for their special action workflow. Would be nice if they made the underlying firewall system and logging of attempts accessible for other actions !
